### PR TITLE
Implement TryFrom for models

### PR DESCRIPTION
**Pull request by heaths**

### CODE REVIEW

**Comment by heaths** [comment]
@JeffreyRichter here's an example of what I was talking about with `Into` - actually, `TryInto` since deserialization can also fail.

**Comment by heaths** [comment]
/cc @jamesbell @demoray @rickwinter for feedback

The discussion Jeffrey and I had was that I believe getting a `T` from a `Response` is far easier and more efficient than getting a `Response` from a `T`. The latter requires some reserved name, though making it a method mitigates any potential name collisions as long as we stick to public model fields for data. But this also lets us deserialize `T` only when necessary; however, it does mean that we need a `Response<T>` so that users have type-safety (and hints, etc.) as to what type a client method returns, unlike the prototypes now in this repo. Alternatively, we could "scope" the method to a trait so that we never have a name collision and we can use whatever private field name e.g., `_raw_response` or even `_{generated GUID}` (I'm kidding...mostly) we want e.g.,

```rust
impl Into<Response> for Secret {
    fn into(self) -> Response {
      self._raw_response
    }
}
```

That said, Jeffrey and I both agree that the mainline scenario is just getting the model; however, there are reasons to want the response yourself and many languages expose that, so I think we should from the start.

**Comment by JeffreyRichter** [comment]
https://github.com/heaths/azure-sdk-for-rust-proto/blob/9c4011f80110636251b4becc7a1d350af8e983e8/sdk/client_new_method_params_context/examples/set_secret_params_context.rs#L12

Some comment we can discuss...
Line 12: Is Arc required or is it optional?
16: We should only support exponential backoff; no other kind.
20: Is None/Some(Xxx) better than Xxx:new() for a default everywhere?
30/31 & 51/52: I would have expected ctx to not be mut and 31 to be ctx = ctx.insert(...)
37-41: A lot of Some() usage. Might this get annoying for customers?

I'm struggling with our mainstream scenario (customer using the deserialized response) requiring an extra step.
That said, I'm willing to get other people's opinion on it and react to feedback.

**Comment by heaths** [comment]
@JeffreyRichter was there some reason you didn't comment on specific lines via GitHub? This format is unproductive - it's very hard to follow and respond to individual issues.

**Comment by heaths** [comment]
> Line 12: Is Arc required or is it optional?

It's required. Credentials can be used in different async pipeline calls.

> We should only support exponential backoff; no other kind.

It's an example. But we do also support other retry policies just like every other language. Exponential isn't always best.

> Is None/Some(Xxx) better than Xxx:new() for a default everywhere?

Why allocate unnecessarily? `Option<T>` is idiomatic and we'll stick with that.

> I would have expected ctx to not be mut and 31 to be ctx = ctx.insert(...)

This prototype does not include the recent `Context` changes I demo in https://gist.github.com/heaths/3aea54ccac2e404d22e110b2400e9200. Again, this is an example. For this PR, please focus on what changes with regard to response models.

> A lot of Some() usage. Might this get annoying for customers?

This is idiomatic to Rust for `Option<T>`. Had we stuck with builders as exist in the repo now, we would need far fewer `Option<T>`. It's no different than Go taking a pointer to every parameter or model field: every caller has to do a little something extra just to pass a parameter. I don't see a difference.

**Comment by heaths** [comment]
@JeffreyRichter @jamesbell @demoray @RickWinter I just added a commit that shows an example of a model containing a raw response. We don't necessarily have to use a trait, but this effectively mitigates a possible name collision.

The main drawback is that we still have to stream all bytes and deserialize into a model whether the customer wants it or not. Yes, most will, but `send(...).await?.into()` could work just as easily. Ideally, we could eliminate the final `into()`.

**Comment by heaths** [comment]
Thinking about `Response<T>` vs. `T` with a raw response, I strongly believe `Response<T>` is the way to go because not every response is JSON - it could also contain XML or even binary. Much like .NET has their `BinaryData`, this provides us an opportunity to have a consistent result type that is flexible enough to handle the different response types. Yes, it's an extra call, but people will be used to doing it.

Comparing with AWS SDK, for example, they do return strongly-typed result models but only expose, by trait impl, their equivalent of our `x-ms-request-id`, which varies for some services. And then what about `x-ms-client-request-id` if provided and supported? What about the gobs of other headers different services pack meaning into that might be interesting for some use cases? As a low-level language, I believe we should expose the entire response. Still, I'm open to better ways. Right now, we have either:

1. (Preferred) A `Response<T>` that would provide methods like `json(self) -> T`, `xml(self) -> T`, and/or `try_into(self) -> T` (where, since we know the format, we just deserialize `T`). We could even look into serde's no-allocation feature, though that means models will have to use `Cow`s. This can deserialize `T` while downloading and consuming bytes from the network stream. There should be little to no duplication of memory.
2. A `T` that is invariably deserialized but retains an in-memory, fully downloaded response, thus effectively duplicating memory for the `T` and cached response.
3. Any other ideas?

/cc @JeffreyRichter @RickWinter @demoray @jamesbell

**Comment by JeffreyRichter** [comment]
I feel like we're still a little incomplete here.
An HTTP method response includes 2 things that are operation specific:

1. The response headers
2. The response body.

I think the body can be represented by a generic T.
But the response headers are not generic and need to be in a method-specific response datatype.
The response of a client method can either be an error (non-2xx status code) which includes the HTTP status code, the Azure error string and the ra JSON body (used for debugging/logging). 

OR, the 2xxx response is a structure with response headers and a T. The response header info is ephemeral meaning that it is usually examined right after the function all and then discarded while the T may live longer (typically for passing around the client app for processing, caching or for modify/update). The raw JSON/XML response is also ephemeral like the HTTP response headers. If the customer wants it, they can use it then and there or we drop it.

I think the above is the best mental model to have when designing the implementation.

**Comment by heaths** [comment]
The `Response` already includes the status code, headers, and body in the existing implementation, upon which I'm building. What is incomplete here? All I'm doing is defining generic parameter `T` on our existing `Response` that lets you deserialize it when needed, or retain the entire response.

See <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=39fcc4ad40e53c2db55b97bc7398ff8d> for an example. The models and `main` function are at the top and everything below is implementation detail - much of which is what we have been using successfully to date.

The client method already returns a `Result<T, E>` where `E` is `azure_core::Error` that contains the error HTTP status code and error message. We *must* also extend this to store the response, which a number of internal and external customers use to get error details.

That said, all the gist and my recommendation do is move the type declaration of `T` from deserialization e.g., `Response::json::<T>()` to `Response<T>` itself so that the client method always defines (as a hint, type safety, etc.) what `T` is. Customers can still ignore that and deserialize into their own models or otherwise parse the response body as needed.

/cc @LarryOsterman

**Comment by JeffreyRichter** [comment]
I see, you have headers as a HashMap<String, String>.
In Go & C++, we have the headers as proper fields with type-safe data types like ETag or time.Time.
So, in Go/C++, each method has its own response datatype with 1 field/datatype per header.

With a hashmap, the customer must manually parse an etag or last-modified header value (just to give some examples). Are you sure you don't want to deserialize response headers into type-safe fields? What about Blob metadata headers? Our Blob SDK would provide to value to assist users here?

**Comment by heaths** [comment]
> I see, you have headers as a HashMap<String, String>. In Go & C++, we have the headers as proper fields with type-safe data types like ETag or time.Time. So, in Go/C++, each method has its own response datatype with 1 field/datatype per header.
> 
> With a hashmap, the customer must manually parse an etag or last-modified header value (just to give some examples). Are you sure you don't want to deserialize response headers into type-safe fields? What about Blob metadata headers? Our Blob SDK would provide to value to assist users here?

It's a simplified prototype. I also don't have a proper HTTP pipeline defined, but this does replicate the most important parts of our pipeline currently.

Our current `Headers` do have defined headers: <https://github.com/Azure/azure-sdk-for-rust/blob/main/sdk/core/src/headers/mod.rs>

The focus of this simplified gist was to show the calling convention I'm recommending to be as versatile as other languages why not having access to a GC to clean up or effectively duplicating memory.
